### ISSUE TITLE
Coroutines scheduler cannot create as many threads as set by kotlinx.coroutines.io.parallelism property

### ISSUE THREAD

**Issue by elizarov** [Kotlin/kotlinx.coroutines#3133]
Reproduces on some machines, this code fails to create 10k threads:

```
import kotlinx.coroutines.*

fun main() {
    System.setProperty(IO_PARALLELISM_PROPERTY_NAME, 10000.toString());
    runBlocking {
        for (i in 1..10000) {
            launch(Dispatchers.IO) {
                while (true) {
                    println("$i: ${Thread.currentThread().name}")
                    Thread.sleep(10000L)
                }
            }
        }
    }
}
```

**Comment by elizarov** [comment]
Ok. This is not a race. It gets limited by `kotlinx.coroutines.scheduler.max.pool.size` which is set to `128 * numberOfCores` by default.

**Comment by elizarov** [comment]
At least, we need to account for "kotlinx.coroutines.io.parallelism" when determining max pool size. Ideally, dynamically growing this limit up to a hardcoded maximum pool size (around 2M threads) should be supported to truly realize the promise of Dispatchers.IO elasticity.

**Comment by elizarov** [comment]
Ok. That seems to affect only pre-1.6.0 versions. It does not reproduce in 1.6.0 and the limit seems to be fixed in 1.6.0 by this commit https://github.com/Kotlin/kotlinx.coroutines/commit/1df0be553c5c0558feeb7299114239442bd3d33c
